### PR TITLE
swiping through images in viewer widget

### PR DESCRIPTION
**Pull request by JakubSerafin**
Viewer widget will show buttons prev and next on top, when more that one url can be found in the cell. 
Also, images can be swiped on mobile via gesture.

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -4,241 +4,269 @@
 
 "@cspotcode/source-map-support@^0.8.0":
   version "0.8.1"
-  resolved "https://registry.yarnpkg.com/@cspotcode/source-map-support/-/source-map-support-0.8.1.tgz#00629c35a688e05a88b1cda684fb9d5e73f000a1"
+  resolved "https://registry.npmjs.org/@cspotcode/source-map-support/-/source-map-support-0.8.1.tgz"
```
Comment (berhalak) [inline yarn.lock:7]: You need to revert this. Normal yarn uses yarnnpkg, and your does not. So the next person (or any auto script) will have problems.

File: test/init-mocha-webdriver.ts [hunk 2]
```diff
@@ -26,6 +26,7 @@ exports.mochaHooks = getMochaHooks();
 
 let server: GristTestServer;
 exports.mochaGlobalSetup = async function() {
+
```
Comment (berhalak) [inline test/init-mocha-webdriver.ts:29]: Not needed

File: package.json [hunk 3]
```diff
@@ -27,9 +27,11 @@
     "node-fetch": "^2",
     "nodemon": "^2.0.15",
     "ts-node": "^10.9.1",
-    "typescript": "^5.1.3"
+    "typescript": "^5.1.3",
+    "chromedriver": "^120.0.0"
```
Comment (berhalak) [inline package.json:31]: Why this is needed?

File: test/fixtures/images/image1.jpg
Comment (berhalak) [inline test/fixtures/images/image1.jpg:1]: Those are some AI generated images (so might cause licencing issues). Better use some Grist images from our site.

File: test/gristWebDriverUtils.ts [hunk 4]
```diff
@@ -287,7 +319,8 @@ export class GristWebDriverUtils {
    */
   public async undo(optCount: number = 1, optTimeout?: number) {
     for (let i = 0; i < optCount; ++i) {
-      await this.driver.find('.test-undo').doClick();
+        const undoButton = await this.driver.find('.test-undo');
```
Comment (berhalak) [inline test/gristWebDriverUtils.ts:322]: Not needed

File: viewer/index.html [hunk 5]
```diff
@@ -27,58 +58,33 @@
...
+        display: flex;
+        flex-direction: column;
+      }
+
     </style>
+    <meta name="viewport" content="user-scalable=no, width=device-width, initial-scale=1, maximum-scale=1">
   </head>
   <body>
+  <div id="navigation-buttons" style="display: none">
+    <button id="calendar-button-previous" class="btn btn-sm" onclick="imageRotator.PreviousImage()">
```
Comment (berhalak) [inline viewer/index.html:79]: `btn-sm` is not defined, is it?

File: viewer/index.html [hunk 6]
```diff
@@ -27,58 +58,33 @@
           height: 100%;
           object-fit: contain;
       }
+      #navigation-buttons{
+        width: 100%;
```
Comment (berhalak) [inline viewer/index.html:62]: `width: 100%` not needed. And as you haven't changed box-model, it can cause bugs latter on.

File: viewer/script.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
+class ImageRotator {
+    _urls = [];
+    _urlIndex = 0;
+
+    PreviousImage() {
```
Comment (berhalak) [inline viewer/script.js:5]: methods use camel case (first letter lowercase).

File: viewer/script.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    setImages(urls) {
+        this._urls = urls;
+        this._urlIndex = 0;
+        this.showImage();
+    }
+
+    showImage() {
+        let url = this._urls[this._urlIndex];
```
Comment (berhalak) [inline viewer/script.js:30]: use `const` here.

File: viewer/script.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        if (!url || url === null || url === undefined) {
```
Comment (berhalak) [inline viewer/script.js:31]: `!url` is enough (it also tests for null or undefined)

File: viewer/script.js [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+class SwipeHandler {
+    _startX
+    _endX
+    //this sets the minimum swipe distance, to avoid noise and to filter actual swipes from just moving fingers
+    _treshold = 100;
+
+//Function to handle swipes
+    _handleTouch() {
+        //calculate the distance on x-axis and o y-axis. Check wheter had the great moving ratio.
+        var xDist = this._endX - this._startX;
```
Comment (berhalak) [inline viewer/script.js:54]: Don't use `var`

File: viewer/script.js [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        }
+    }
+
+    constructor(element) {
+        this.viewer = element;
+    }
+}
+
+class SwipeHandler {
+    _startX
```
Comment (berhalak) [inline viewer/script.js:46]: `;` is missing.

File: viewer/script.js [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        if (Math.abs(xDist) > this._treshold) {
+            if (xDist > 0) {
+                this.onSwipeRight();
+            } else {
+                this.onSwipeLeft();
+            }
+        }
+    }
+
+    onSwipeLeft
```
Comment (berhalak) [inline viewer/script.js:64]: Should be at the beggining and have `;` at the end.

File: viewer/script.js [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    onSwipeRight
+
+    constructor() {
+        let viewer = document.getElementById('viewer');
```
Comment (berhalak) [inline viewer/script.js:68]: `const` instead of `let`

File: viewer/script.js [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        viewer.addEventListener('touchstart', (event) => {
+            this._startX = event.touches[0].clientX;
+
```
Comment (berhalak) [inline viewer/script.js:71]: Extra lines

File: viewer/script.js [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+            this._handleTouch()
+
+        })
+    }
+}
+
+let imageRotator
+let swipeHandler
+window.onload = function () {
+    imageRotator = new ImageRotator(document.getElementById('viewer'));
```
Comment (berhalak) [inline viewer/script.js:84]: Since constructor doesn't do anything (which is good), you can just pass the selector there (`viewer`) and left the `ImageRotator` instance the job for finding it when its needed. That way, you won't need the `window.onload` event handler.

File: viewer/script.js [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+    swipeHandler = new SwipeHandler();
+    swipeHandler.onSwipeLeft = () => imageRotator.NextImage();
+    swipeHandler.onSwipeRight = () => imageRotator.PreviousImage();
+}
+
+function showError(msg) {
+    var el = document.getElementById('error')
```
Comment (berhalak) [inline viewer/script.js:91]: `var`

File: viewer/script.js [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+    if (!msg) {
+        el.style.display = 'none';
+    } else {
+        el.innerHTML = msg;
+        el.style.display = 'block';
+    }
+}
+
+function toggleNavigationButtons(show = false) {
+    let buttons = document.getElementById('navigation-buttons');
```
Comment (berhalak) [inline viewer/script.js:101]: `const`

File: viewer/script.js (hunk 10 above)
Comment (berhalak) [reply viewer/script.js:54]: Indention is wrong in the whole file.

File: viewer/script.js [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        this.showImage();
+    }
+
+    showImage() {
+        let url = this._urls[this._urlIndex];
+        if (!url || url === null || url === undefined) {
+            this.viewer.style.display = 'none';
+        } else {
+            this.viewer.src = url;
+            this.viewer.alt = `URL: ${url}`; // Set alt parameter to show the URL of the picture
```
Comment (berhalak) [inline viewer/script.js:35]: Useless comment.

File: viewer/script.js [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        if (!data) {
+            viewer.style.display = 'none';
+        } else {
+            const imageUrlRegex = /(https?:\/\/[^\s]+)/g;
+            urls = data.match(imageUrlRegex) || [];
+            toggleNavigationButtons(urls.length > 1)
+            imageRotator.setImages(urls)
+        }
+    }
+});
```
Comment (berhalak) [inline viewer/script.js:145]: No new line at the end.

**Review by berhalak — CHANGES_REQUESTED** [review]
It is good starting point, things that should be improved (in my opinion)
- Overall UI, currently it is not pretty, we can easly use any out of box MIT library https://linxtion.com/demo/react-image-gallery/
- The size of the window is wrong, I see scrollbars all the time.

**Comment by JakubSerafin** [comment]
> It is good starting point, things that should be improved (in my opinion)
> 
>     * Overall UI, currently it is not pretty, we can easly use any out of box MIT library https://linxtion.com/demo/react-image-gallery/
> 
>     * The size of the window is wrong, I see scrollbars all the time.

Overall UI improved to be more similar to that in linxtion. 
Cannot spot any scrollbar in chrome/firefox unix chrome, but added overlay handling anyway

**Comment by berhalak** [comment]
When you switch rows from a row that has multiple images to a row with no images at all, those arrows are still there.

File: test/init-mocha-webdriver.ts [hunk 20]
```diff
@@ -25,7 +25,7 @@ if (process.env.MOCHA_WEBDRIVER_NO_CONTROL_BANNER === undefined) {
 exports.mochaHooks = getMochaHooks();
 
 let server: GristTestServer;
-exports.mochaGlobalSetup = async function() {
+exports.mochaGlobalSetup = async function() {
```
Comment (berhalak) [inline test/init-mocha-webdriver.ts:28]:
```suggestion
exports.mochaGlobalSetup = async function() {
```

File: test/gristWebDriverUtils.ts [hunk 21]
```diff
@@ -306,6 +338,103 @@ export class GristWebDriverUtils {
...
+  }
+
+
+  /**
+   * Returns a visible GridView cell. Options may be given as arguments directly, or as an object.
+   * - col: column name, or 0-based column index
+   * - rowNum: 1-based row numbers, as visible in the row headers on the left of the grid.
+   * - section: optional name of the section to use; will use active section if omitted.
+   */
+  getCell(col: number|string, rowNum: number, section?: string): WebElementPromise;
```
Comment (berhalak) [inline test/gristWebDriverUtils.ts:375]: Some methods are missing `public` modifier

File: test/viewer.ts [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+  async function swipeLeft(element: string | RegExp) {
+    grist.executeScriptInCustomWidget(swipeScript, element,'left')
+  }
+
+  async function swipeRight(element: string | RegExp) {
+    grist.executeScriptInCustomWidget(swipeScript, element,'right')
+  }
+
+  function swipeScript(elementTag: string , direction: 'left' | 'right') {
+      let mode = direction;
```
Comment (berhalak) [inline test/viewer.ts:173]: There is some weird indention in some functions, for example here.

File: viewer/index.html [hunk 23]
```diff
@@ -1,84 +1,116 @@
...
+    <button id="calendar-button-next" class="btn" onclick="imageRotator.nextImage()">
+      <span class="icon icon-arrow-right"></span>
+    </button>
+  </div>
+  <div id="error"></div>
+  <img id="viewer" alt="image loaded by gris from table" src="" />
+  <script src="script.js"></script>
+</body>
+
+</html>
```
Comment (berhalak) [inline viewer/index.html:116]: Space at the end of the file.

File: test/gristWebDriverUtils.ts [hunk 24]
```diff
@@ -359,10 +362,16 @@ public async sendCommand(name: string, argument: any = null) {
     await this.waitForServer();
   }
 
-  public async focusOnWidget(widgetName: string|RegExp) {
-      await this.driver.findContentWait('.test-widget-title-text', widgetName, 5000).click();
-      await this.driver.findWait(".test-widget-title-cancel", 1000).click();
-      await this.waitForServer();
+  public async focusOnWidget(widgetName: string | RegExp) {
```
Comment (berhalak) [inline test/gristWebDriverUtils.ts:365]: What is this? It looks wrong. `findContent` will throw error if element is not present I think.
What does ctrl+O does? If it is correct it should also work on macs (with mod key).

File: viewer/index.html [hunk 25]
```diff
@@ -1,84 +1,117 @@
...
-    </script>
-  </body>
+
+<head>
+  <meta charset="utf-8">
+  <title>A cheap and cheerful image viewer</title>
+  <script src="https://docs.getgrist.com/grist-plugin-api.js"></script>
+  <style>
+    .icon-arrow-left {
+      mask-image: var(--icon-ArrowLeft);
```
Comment (berhalak) [inline viewer/index.html:10]: I think that adding a hover cursor, to indicate that this is clickable element, would be useful.

File: test/gristWebDriverUtils.ts (hunk 24 above)
Comment (JakubSerafin) [reply test/gristWebDriverUtils.ts:365]: `isPresent()` should just return false is element is not present. 
CTRL+O is shortcut to focus on next widget.

**Review by berhalak — APPROVED** [review]
Thanks
